language/move-stdlib/nursery/Vault.move

```diff
@@ -0,0 +1,462 @@
+        // We (should be/are) able to prove that the below cannot happen, but we leave the assertion
+        // here anyway for double safety.
+        assert(Option::is_none(content), Errors::internal(EACCESSOR_INCONSISTENCY));
+        Option::fill(content, new_content);
+    }
+
+
+    // ================================================================================================================
+    // Delegation
+
+    /// Delegates the right to acquire a capability of the given type.
+    public fun delegate<Content: store + drop>(cap: &DelegateCap<Content>, to_signer: &signer, cap_type: CapType)
+    acquires Vault, VaultDelegate {
+        assert(
+            borrow_global<Vault<Content>>(cap.vault_address).supports_delegation,
+            Errors::invalid_state(EDELEGATION_NOT_SUPPORTED)
+        );
+
+        let addr = Signer::address_of(to_signer);
+        assert(addr != cap.vault_address, Errors::invalid_argument(EDELEGATE_TO_SELF));
+
+        if (!exists<VaultDelegate<Content>>(addr)) {
+            // Create VaultDelegate if it is not yet existing.
+            move_to<VaultDelegate<Content>>(
+                to_signer,
+                VaultDelegate{vault_address: cap.vault_address, granted_caps: Vector::empty()}
+            );
+            let vault = borrow_global_mut<Vault<Content>>(cap.vault_address);
+            add_element(&mut vault.delegates, addr);
+        };
+        let delegate = borrow_global_mut<VaultDelegate<Content>>(addr);
+        add_element(&mut delegate.granted_caps, *&cap_type);
+
+        // Generate event
+        emit_delegate_event(cap, cap_type, addr, false);
+    }
+
+    /// Revokes the delegated right to acquire a capability of given type.
+    public fun revoke<Content: store + drop>(cap: &DelegateCap<Content>, addr: address, cap_type: CapType)
+    acquires Vault, VaultDelegate {
+        assert(exists<VaultDelegate<Content>>(addr), Errors::not_published(EDELEGATE));
+        let delegate = borrow_global_mut<VaultDelegate<Content>>(addr);
+        remove_element(&mut delegate.granted_caps, &cap_type);
+        // If the granted caps of this delegate drop to zero, remove it.
+        if (Vector::is_empty(&delegate.granted_caps)) {
+            let VaultDelegate{ vault_address: _owner, granted_caps: _granted_caps} =
+                move_from<VaultDelegate<Content>>(addr);
+            let vault = borrow_global_mut<Vault<Content>>(cap.vault_address);
+            remove_element(&mut vault.delegates, &addr);
+        };
+
+        // Generate event.
+        emit_delegate_event(cap, cap_type, addr, true);
+    }
+
+    /// Revokes all delegate rights for this vault.
+    public fun revoke_all<Content: store + drop>(cap: &DelegateCap<Content>)
+    acquires Vault, VaultDelegate {
+        while (!Vector::is_empty(&borrow_global<Vault<Content>>(cap.vault_address).delegates)) {
+            let addr = Vector::pop_back(&mut borrow_global_mut<Vault<Content>>(cap.vault_address).delegates);
```

Comment:
Can these borrow_globals be pulled outside this loop?

Reply:
Unfortunately not because the emit_delegate_event acquires the resource.